source/h1_connection.c

```diff
@@ -721,6 +722,74 @@ static void s_client_update_incoming_stream_ptr(struct aws_h1_connection *connec
     s_set_incoming_stream_ptr(connection, desired);
 }
 
+static void s_http_request_idle_timeout_task(struct aws_task *task, void *arg, enum aws_task_status status) {
+    struct aws_h1_stream *stream = arg;
+    if (status == AWS_TASK_STATUS_CANCELED) {
+        goto cleanup;
+    }
+    struct aws_http_connection *connection_base = stream->base.owning_connection;
+    struct aws_h1_connection *connection = AWS_CONTAINER_OF(connection_base, struct aws_h1_connection, base);
+    /* Time out happened, close the connection manually */
+    uint64_t idle_timeout_ms = stream->base.client_data->idle_timeout_ms == 0
+                                   ? connection_base->client_data->idle_timeout_ms
+                                   : stream->base.client_data->idle_timeout_ms;
+    AWS_LOGF_INFO(
+        AWS_LS_HTTP_CONNECTION,
+        "id=%p: Closing connection as timeout after request sent to the first byte received happened. Idle time "
+        "is %" PRIu64 " ms",
+        (void *)connection_base,
+        idle_timeout_ms);
+
+    /* Don't stop reading/writing immediately, let that happen naturally during the channel shutdown process. */
+    s_stop(
+        connection,
+        false /*stop_reading*/,
+        false /*stop_writing*/,
+        true /*schedule_shutdown*/,
+        AWS_ERROR_HTTP_REQUEST_IDLE_TIMEOUT);
+    connection_base->client_data->request_idle_timeout_task = NULL;
+
+cleanup:
+    aws_mem_release(stream->base.alloc, task);
+}
+
+static void s_request_finished_sending(struct aws_h1_stream *stream) {
+    struct aws_http_connection *connection = stream->base.owning_connection;
+    struct aws_channel *channel = aws_http_connection_get_channel(connection);
+    AWS_ASSERT(aws_channel_thread_is_callers_thread(channel));
+
+    stream->is_outgoing_message_done = true;
```

Comment:
In the current flow, I'm not 100% sure whether this function might end up being called twice.
probably want to check whether `is_outgoing_message_done` is already true